File: Rules.cs
#nullable disable
using System.Globalization;
using System.Text.RegularExpressions;

public class Violation
{
    public string Rule, Severity, Message, Suggestion;
    public int Line;

    public Violation(string rule, string severity, string message, string suggestion, int line)
    {
        Rule = rule; Severity = severity; Message = message; Suggestion = suggestion; Line = line;
    }
}

public static class Rules
{
    static readonly string[] AccessMods = { "public", "private", "protected", "internal" };
    static readonly string[] UnityLifecycle = {
        "Update", "FixedUpdate", "LateUpdate", "OnEnable", "OnDisable", "OnDestroy",
        "OnCollisionEnter", "OnCollisionStay", "OnCollisionExit",
        "OnTriggerEnter", "OnTriggerStay", "OnTriggerExit"
    };
    static readonly Regex NumberRe = new Regex(@"\b\d+(?:\.\d+)?[fFdDmMuUlL]?\b");
    static readonly Regex EnumRe = new Regex(@"^\w+\s*(?:=\s*[\d.]+[fFdDmMuUlL]?\s*)?[,]?$");
    static readonly double[] SafeNums = { 0, 1, 2 };
    static readonly string[] DeclWords = { "public", "private", "protected", "internal", "readonly", "static", "const" };

    static Violation V(string rule, string sev, string msg, string hint, int line) =>
        new Violation(rule, sev, msg, hint, line);

    static bool IsPascal(string n) => n.Length > 0 && char.IsUpper(n[0]) && !n.Contains('_');
    static bool IsCamel(string n) => n.Length > 0 && char.IsLower(n[0]) && !n.Contains('_');
    static bool IsValidField(string n) => n.StartsWith("_") && n.Length > 1 && char.IsLower(n[1]) && !n.Substring(1).Contains('_');
    static string ToPascal(string n) => n.Length == 0 ? n : char.ToUpper(n[0]) + n.Substring(1);
    static string ToCamel(string n) { var s = n.TrimStart('_'); return s.Length == 0 ? s : char.ToLower(s[0]) + s.Substring(1); }
    static bool HasAccessMod(List<string> mods) => mods.Any(m => AccessMods.Contains(m));

    public static List<Violation> Run(List<Declaration> decls, string source, bool useLlm)
    {
        var violations = new List<Violation>();
        violations.AddRange(CheckNoPublicFields(decls));
        violations.AddRange(CheckAccessModifiers(decls));
        violations.AddRange(CheckTypeNames(decls));
        violations.AddRange(CheckMethodNames(decls));
        violations.AddRange(CheckVariableNames(decls));
        violations.AddRange(CheckPrivateFieldNames(decls));
        violations.AddRange(CheckMagicNumbers(source));
        violations.AddRange(CheckSerializeField(decls));
        violations.AddRange(CheckAwakeVsStart(decls));
        if (useLlm) violations.AddRange(CheckNamesWithLlm(decls));
        violations.Sort((a, b) => a.Line - b.Line);
        return violations;
    }

    static List<Violation> CheckNoPublicFields(List<Declaration> decls)
    {
        var violations = new List<Violation>();
        foreach (var d in decls)
        {
            if (d.Kind != "field" || !d.Modifiers.Contains("public") || d.Modifiers.Contains("const")
                || (d.Modifiers.Contains("static") && d.Modifiers.Contains("readonly"))) continue;
            violations.Add(V("CQE001", "error", $"Public field '{d.Name}' should be a property.",
                $"Replace with: public Type {ToPascal(d.Name)} {{ get; set; }}", d.Line));
        }
        return violations;
    }

    static List<Violation> CheckAccessModifiers(List<Declaration> decls)
    {
        var violations = new List<Violation>();
        foreach (var d in decls)
        {
            if (d.Kind == "variable" || HasAccessMod(d.Modifiers)) continue;
            violations.Add(V("CQE002", "error",
                $"{char.ToUpper(d.Kind[0]) + d.Kind.Substring(1)} '{d.Name}' has no access modifier.",
                $"Add 'private' (or another modifier) before '{d.Name}'.", d.Line));
        }
        return violations;
    }

    static List<Violation> CheckTypeNames(List<Declaration> decls)
    {
        var violations = new List<Violation>();
        foreach (var d in decls)
        {
            if (d.Kind is not ("class" or "struct" or "interface" or "enum" or "record")) continue;
            if (!IsPascal(d.Name))
                violations.Add(V("CQE003", "error", $"Type '{d.Name}' is not PascalCase.", $"Rename to '{ToPascal(d.Name)}'.", d.Line));
        }
        return violations;
    }

    static List<Violation> CheckMethodNames(List<Declaration> decls)
    {
        var violations = new List<Violation>();
        foreach (var d in decls)
            if (d.Kind == "method" && !IsPascal(d.Name))
                violations.Add(V("CQE004", "error", $"Method '{d.Name}' is not PascalCase.", $"Rename to '{ToPascal(d.Name)}'.", d.Line));
        return violations;
    }

    static List<Violation> CheckVariableNames(List<Declaration> decls)
    {
        var violations = new List<Violation>();
        foreach (var d in decls)
            if (d.Kind == "variable" && !IsCamel(d.Name))
                violations.Add(V("CQE005", "error", $"Variable '{d.Name}' is not camelCase.", $"Rename to '{ToCamel(d.Name)}'.", d.Line));
        return violations;
    }

    static List<Violation> CheckPrivateFieldNames(List<Declaration> decls)
    {
        var violations = new List<Violation>();
        foreach (var d in decls)
        {
            if (d.Kind != "field") continue;
            if (d.Modifiers.Contains("const") || d.Modifiers.Contains("static")) continue;
            bool isPrivate = d.Modifiers.Contains("private") || !HasAccessMod(d.Modifiers);
            if (!isPrivate || IsValidField(d.Name)) continue;
            violations.Add(V("CQE006", "error", $"Private field '{d.Name}' should be '_camelCase'.",
                $"Rename to '_{ToCamel(d.Name)}'.", d.Line));
        }
        return violations;
    }

    static List<Violation> CheckMagicNumbers(string source)
    {
        var violations = new List<Violation>();
        int lineNum = 0;
        foreach (var rawLine in source.Split('\n'))
        {
            lineNum++;
            var line = rawLine.Trim();
            if (string.IsNullOrEmpty(line)) continue;
            if (line.Split(' ').Any(w => DeclWords.Contains(w))) continue;
            if (EnumRe.IsMatch(line)) continue;
            foreach (Match m in NumberRe.Matches(line))
            {
                var numStr = m.Value.TrimEnd('f', 'F', 'd', 'D', 'm', 'M', 'u', 'U', 'l', 'L');
                if (!double.TryParse(numStr, NumberStyles.Any, CultureInfo.InvariantCulture, out double val)) continue;
                if (SafeNums.Contains(val)) continue;
                violations.Add(V("CQE008", "warning", $"Magic number '{m.Value}' found.",
                    "Replace with a named constant: const float NAME = value;", lineNum));
            }
        }
        return violations;
    }

    static List<Violation> CheckSerializeField(List<Declaration> decls)
    {
        var violations = new List<Violation>();
        foreach (var d in decls)
        {
            if (d.Kind != "field" || !d.Modifiers.Contains("public") || d.Modifiers.Contains("const")
                || d.Modifiers.Contains("static") || d.Attributes.Contains("SerializeField")) continue;
            violations.Add(V("CQE009", "warning", $"Public field '{d.Name}' exposes Unity serialized data.",
                "Use '[SerializeField] private' instead of 'public'.", d.Line));
        }
        return violations;
    }

    static List<Violation> CheckAwakeVsStart(List<Declaration> decls)
    {
        var violations = new List<Violation>();
        var byClass = new Dictionary<string, List<Declaration>>();
        foreach (var d in decls.Where(d => d.Kind == "method"))
        {
            var cls = d.Parent ?? "__global__";
            if (!byClass.ContainsKey(cls)) byClass[cls] = new List<Declaration>();
            byClass[cls].Add(d);
        }
        foreach (var kv in byClass)
        {
            var methods = kv.Value;
            bool hasAwake = methods.Any(m => m.Name == "Awake");
            bool hasStart = methods.Any(m => m.Name == "Start");
            bool hasLifecycle = methods.Any(m => UnityLifecycle.Contains(m.Name));
            switch (hasAwake, hasStart, hasLifecycle)
            {
                case (true, true, _):
                    var awake = methods.First(m => m.Name == "Awake");
                    violations.Add(V("CQE010", "warning", $"'{kv.Key}' has both Awake() and Start().",
                        "Consolidate initialization into Awake() only.", awake.Line));
                    break;
                case (false, false, true):
                    var first = methods.Where(m => UnityLifecycle.Contains(m.Name)).OrderBy(m => m.Line).First();
                    violations.Add(V("CQE010", "warning", $"'{kv.Key}' has Unity callbacks but no Awake() or Start().",
                        "Add an Awake() method for explicit initialization.", first.Line));
                    break;
            }
        }
        return violations;
    }

    static List<Violation> CheckNamesWithLlm(List<Declaration> decls)
    {
        var candidates = decls.Where(d => d.Kind != "variable").Select(d => new NameKind(d.Name, d.Kind)).ToList();
        var flagged = Llm.FindBadNames(candidates);
        var linePerName = new Dictionary<string, int>();
        foreach (var d in decls) linePerName[d.Name] = d.Line;
        var violations = new List<Violation>();
        foreach (var r in flagged)
        {
            linePerName.TryGetValue(r.Name, out int line);
            violations.Add(V("CQE011", "warning", $"Unclear name '{r.Name}': {r.Reason}",
                "Use a clear, descriptive English identifier.", line));
        }
        return violations;
    }
}

File: src/rules.py
"""
All code convention rules.

Each rule is a plain function that receives the declarations list (and sometimes
the cleaned source) and returns a list of violation dicts.

run_rules() calls them all and returns violations sorted by line number.
"""

import re
from .parser import Declaration

ACCESS_MODIFIERS = {"public", "private", "protected", "internal"}

UNITY_LIFECYCLE = {
    "Update", "FixedUpdate", "LateUpdate",
    "OnEnable", "OnDisable", "OnDestroy",
    "OnCollisionEnter", "OnCollisionStay", "OnCollisionExit",
    "OnTriggerEnter", "OnTriggerStay", "OnTriggerExit",
}


# ── Helpers ────────────────────────────────────────────────────────────────────

def _violation(rule: str, severity: str, message: str, suggestion: str, line: int) -> dict:
    return {"rule": rule, "severity": severity, "message": message,
            "suggestion": suggestion, "line": line}


def _to_pascal(name: str) -> str:
    return name[0].upper() + name[1:] if name else name


def _to_camel(name: str) -> str:
    clean = name.lstrip("_")
    return clean[0].lower() + clean[1:] if clean else name


def _is_pascal(name: str) -> bool:
    return bool(name) and name[0].isupper() and "_" not in name


def _is_camel(name: str) -> bool:
    return bool(name) and name[0].islower() and "_" not in name


def _is_private_field_format(name: str) -> bool:
    """_camelCase: starts with _, second char lowercase, no further underscores."""
    return (
        name.startswith("_")
        and len(name) > 1
        and name[1].islower()
        and "_" not in name[1:]
    )


# ── CQE001: no public fields ───────────────────────────────────────────────────

def check_no_public_fields(declarations: list[Declaration]) -> list[dict]:
    """Public fields should be properties. Exempt: const and static readonly."""
    violations = []
    for decl in declarations:
        if decl.kind != "field":
            continue
        if "public" not in decl.modifiers:
            continue
        if "const" in decl.modifiers:
            continue
        if "static" in decl.modifiers and "readonly" in decl.modifiers:
            continue
        violations.append(_violation(
            rule="CQE001", severity="error",
            message=f"Public field '{decl.name}' should be a property.",
            suggestion=f"Replace with: public Type {_to_pascal(decl.name)} {{ get; set; }}",
            line=decl.line,
        ))
    return violations


# ── CQE002: access modifier required ──────────────────────────────────────────

def check_access_modifiers(declarations: list[Declaration]) -> list[dict]:
    """Every class, method, and field must have an explicit access modifier."""
    violations = []
    for decl in declarations:
        if decl.kind == "variable":
            continue  # local variables don't have access modifiers
        if ACCESS_MODIFIERS & set(decl.modifiers):
            continue
        violations.append(_violation(
            rule="CQE002", severity="error",
            message=f"{decl.kind.capitalize()} '{decl.name}' has no access modifier.",
            suggestion=f"Add 'private' (or another modifier) before '{decl.name}'.",
            line=decl.line,
        ))
    return violations


# ── CQE003: class names PascalCase ────────────────────────────────────────────

def check_type_names(declarations: list[Declaration]) -> list[dict]:
    """Type names (class, struct, interface, enum, record) must be PascalCase."""
    violations = []
    for decl in declarations:
        if decl.kind not in ("class", "struct", "interface", "enum", "record"):
            continue
        if _is_pascal(decl.name):
            continue
        violations.append(_violation(
            rule="CQE003", severity="error",
            message=f"Type '{decl.name}' is not PascalCase.",
            suggestion=f"Rename to '{_to_pascal(decl.name)}'.",
            line=decl.line,
        ))
    return violations


# ── CQE004: method names PascalCase ───────────────────────────────────────────

def check_method_names(declarations: list[Declaration]) -> list[dict]:
    """Method names must be PascalCase."""
    violations = []
    for decl in declarations:
        if decl.kind != "method":
            continue
        if _is_pascal(decl.name):
            continue
        violations.append(_violation(
            rule="CQE004", severity="error",
            message=f"Method '{decl.name}' is not PascalCase.",
            suggestion=f"Rename to '{_to_pascal(decl.name)}'.",
            line=decl.line,
        ))
    return violations


# ── CQE005: variable names camelCase ──────────────────────────────────────────

def check_variable_names(declarations: list[Declaration]) -> list[dict]:
    """Local variable names must be camelCase."""
    violations = []
    for decl in declarations:
        if decl.kind != "variable":
            continue
        if _is_camel(decl.name):
            continue
        violations.append(_violation(
            rule="CQE005", severity="error",
            message=f"Variable '{decl.name}' is not camelCase.",
            suggestion=f"Rename to '{_to_camel(decl.name)}'.",
            line=decl.line,
        ))
    return violations


# ── CQE006: private fields _camelCase ─────────────────────────────────────────

def check_private_field_names(declarations: list[Declaration]) -> list[dict]:
    """Private fields must follow the _camelCase convention."""
    violations = []
    for decl in declarations:
        if decl.kind != "field":
            continue
        if "const" in decl.modifiers or "static" in decl.modifiers:
            continue
        is_private = (
            "private" in decl.modifiers
            or not (ACCESS_MODIFIERS & set(decl.modifiers))
        )
        if not is_private:
            continue
        if _is_private_field_format(decl.name):
            continue
        violations.append(_violation(
            rule="CQE006", severity="error",
            message=f"Private field '{decl.name}' should be '_camelCase'.",
            suggestion=f"Rename to '_{_to_camel(decl.name)}'.",
            line=decl.line,
        ))
    return violations


# ── CQE008: no magic numbers ──────────────────────────────────────────────────

_NUMBER_RE    = re.compile(r"\b\d+(?:\.\d+)?[fFdDmMuUlL]?\b")
_MAGIC_WHITELIST = frozenset({0, 1, 2})

# Matches bare enum value lines like "Active = 10," or "Inactive,"
_ENUM_VALUE_RE = re.compile(r"^\w+\s*(?:=\s*[\d.]+[fFdDmMuUlL]?\s*)?[,]?$")


def check_magic_numbers(source: str) -> list[dict]:
    """Numeric literals that are not 0, 1 or 2 must be named constants."""
    violations = []
    for lineno, line in enumerate(source.splitlines(), start=1):
        stripped = line.strip()

        if not stripped:
            continue

        words = stripped.split()

        # Skip const declarations and field/member initializations with modifiers
        has_const    = "const" in words
        has_modifier = bool({"public", "private", "protected", "internal",
                             "readonly", "static"} & set(words))
        if has_const or has_modifier:
            continue

        # Skip enum value assignments like "Active = 10," or "Deleted = 30"
        if _ENUM_VALUE_RE.fullmatch(stripped):
            continue

        for match in _NUMBER_RE.finditer(stripped):
            raw = match.group(0).rstrip("fFdDmMuUlL")
            try:
                value = float(raw)
            except ValueError:
                continue
            if value in _MAGIC_WHITELIST:
                continue
            violations.append(_violation(
                rule="CQE008", severity="warning",
                message=f"Magic number '{match.group(0)}' found.",
                suggestion="Replace with a named constant: const float NAME = value;",
                line=lineno,
            ))
    return violations


# ── CQE009: Unity SerializeField ──────────────────────────────────────────────

def check_serialize_field(declarations: list[Declaration]) -> list[dict]:
    """Public fields in Unity should use [SerializeField] private instead."""
    violations = []
    for decl in declarations:
        if decl.kind != "field":
            continue
        if "public" not in decl.modifiers:
            continue
        if "const" in decl.modifiers or "static" in decl.modifiers:
            continue
        if "SerializeField" in decl.attributes:
            continue
        violations.append(_violation(
            rule="CQE009", severity="warning",
            message=f"Public field '{decl.name}' exposes Unity serialized data.",
            suggestion="Use '[SerializeField] private' instead of 'public'.",
            line=decl.line,
        ))
    return violations


# ── CQE010: Awake vs Start ────────────────────────────────────────────────────

def check_awake_vs_start(declarations: list[Declaration]) -> list[dict]:
    """Unity classes should not have both Awake() and Start(), and should have
    at least one of them if they use lifecycle callbacks."""
    violations = []
    methods = [d for d in declarations if d.kind == "method"]

    # Group methods by their parent class
    by_class: dict[str, list[Declaration]] = {}
    for method in methods:
        key = method.parent or "__global__"
        by_class.setdefault(key, []).append(method)

    for class_name, class_methods in by_class.items():
        names = {m.name for m in class_methods}
        has_awake     = "Awake" in names
        has_start     = "Start" in names
        has_lifecycle = bool(names & UNITY_LIFECYCLE)

        if has_awake and has_start:
            awake = next(m for m in class_methods if m.name == "Awake")
            violations.append(_violation(
                rule="CQE010", severity="warning",
                message=f"'{class_name}' has both Awake() and Start().",
                suggestion="Consolidate initialization into Awake() only.",
                line=awake.line,
            ))
        elif has_lifecycle and not has_awake and not has_start:
            first = min((m for m in class_methods if m.name in UNITY_LIFECYCLE),
                        key=lambda m: m.line)
            violations.append(_violation(
                rule="CQE010", severity="warning",
                message=f"'{class_name}' has Unity callbacks but no Awake() or Start().",
                suggestion="Add an Awake() method for explicit initialization.",
                line=first.line,
            ))

    return violations


# ── CQE011: LLM naming quality ────────────────────────────────────────────────

def check_names_with_llm(declarations: list[Declaration]) -> list[dict]:
    """Ask the LLM to identify meaningless or non-English identifier names."""
    try:
        from .llm import find_bad_names
    except ImportError:
        return []

    # Check classes, methods and fields – not local variables (too noisy)
    candidates = [(d.name, d.kind) for d in declarations
                  if d.kind not in ("variable",)]

    bad = find_bad_names(candidates)

    name_to_line = {d.name: d.line for d in declarations}
    violations = []
    for name, reason in bad:
        violations.append(_violation(
            rule="CQE011", severity="warning",
            message=f"Unclear name '{name}': {reason}",
            suggestion="Use a clear, descriptive English identifier.",
            line=name_to_line.get(name, 0),
        ))
    return violations


# ── run_rules ─────────────────────────────────────────────────────────────────

def run_rules(declarations: list[Declaration], source: str,
              use_llm: bool = True) -> list[dict]:
    violations = []
    violations += check_no_public_fields(declarations)
    violations += check_access_modifiers(declarations)
    violations += check_type_names(declarations)
    violations += check_method_names(declarations)
    violations += check_variable_names(declarations)
    violations += check_private_field_names(declarations)
    violations += check_magic_numbers(source)
    violations += check_serialize_field(declarations)
    violations += check_awake_vs_start(declarations)

    if use_llm:
        violations += check_names_with_llm(declarations)

    return sorted(violations, key=lambda v: v["line"])
